File: sdk/clojure/src/rockbox/bluetooth.clj
(ns rockbox.bluetooth
  "Bluetooth devices (Linux only — wraps BlueZ via D-Bus)."
  (:require [rockbox.transport :as t]))

(def ^:private device-fields
  "fragment BluetoothDeviceFields on BluetoothDevice {
     address name paired trusted connected rssi
   }")

(defn devices
  "List paired/known Bluetooth devices."
  [client]
  (:bluetooth-devices
   (t/execute client (str device-fields
                          " query BluetoothDevices { bluetoothDevices { ...BluetoothDeviceFields } }"))))

(defn scan
  "Scan for nearby Bluetooth devices. Optional `timeout-secs`."
  ([client] (scan client nil))
  ([client timeout-secs]
   (:bluetooth-scan
    (t/execute client
               (str device-fields
                    " mutation BluetoothScan($timeoutSecs: Int) {
                        bluetoothScan(timeoutSecs: $timeoutSecs) { ...BluetoothDeviceFields }
                      }")
               {:timeout-secs timeout-secs}))))

(defn connect
  "Connect to a Bluetooth device by address."
  [client address]
  (t/execute client
             "mutation BluetoothConnect($address: String!) { bluetoothConnect(address: $address) }"
             {:address address})
  client)

(defn disconnect
  "Disconnect a Bluetooth device by address."
  [client address]
  (t/execute client
             "mutation BluetoothDisconnect($address: String!) { bluetoothDisconnect(address: $address) }"
             {:address address})
  client)

File: sdk/elixir/lib/rockbox/bluetooth.ex
defmodule Rockbox.Bluetooth do
  @moduledoc """
  Bluetooth device management (Linux only — backed by BlueZ).

  Calls return `{:error, %Rockbox.GraphQLError{}}` on non-Linux hosts with the
  message `"Bluetooth is only supported on Linux"`.
  """

  alias Rockbox.{BluetoothDevice, Client, Transport, Util}

  @fragment ~S"""
    fragment BluetoothDeviceFields on BluetoothDevice {
      address name paired trusted connected rssi
    }
  """

  @spec devices(Client.t()) :: {:ok, [BluetoothDevice.t()]} | {:error, Exception.t()}
  def devices(client) do
    query =
      @fragment <> "query BluetoothDevices { bluetoothDevices { ...BluetoothDeviceFields } }"

    with {:ok, %{"bluetoothDevices" => list}} <- Transport.execute(client, query) do
      {:ok, Util.to_struct_list(BluetoothDevice, list)}
    end
  end

  @spec devices!(Client.t()) :: [BluetoothDevice.t()]
  def devices!(client), do: bang(devices(client))

  @doc "Scan for nearby devices. `timeout_secs` defaults to the firmware default."
  @spec scan(Client.t(), pos_integer() | nil) ::
          {:ok, [BluetoothDevice.t()]} | {:error, Exception.t()}
  def scan(client, timeout_secs \\ nil) do
    query =
      @fragment <>
        "mutation BluetoothScan($timeoutSecs: Int) { bluetoothScan(timeoutSecs: $timeoutSecs) { ...BluetoothDeviceFields } }"

    with {:ok, %{"bluetoothScan" => list}} <-
           Transport.execute(client, query, %{timeout_secs: timeout_secs}) do
      {:ok, Util.to_struct_list(BluetoothDevice, list)}
    end
  end

  @spec connect(Client.t(), String.t()) :: :ok | {:error, Exception.t()}
  def connect(client, address),
    do:
      void(
        Transport.execute(
          client,
          "mutation BluetoothConnect($address: String!) { bluetoothConnect(address: $address) }",
          %{address: address}
        )
      )

  @spec disconnect(Client.t(), String.t()) :: :ok | {:error, Exception.t()}
  def disconnect(client, address),
    do:
      void(
        Transport.execute(
          client,
          "mutation BluetoothDisconnect($address: String!) { bluetoothDisconnect(address: $address) }",
          %{address: address}
        )
      )

  defp void({:ok, _}), do: :ok
  defp void(err), do: err

  defp bang({:ok, value}), do: value
  defp bang({:error, exception}), do: raise(exception)
end
